src/Raven.Server/Documents/DocumentDatabase.cs

```diff
@@ -618,25 +630,37 @@ private void ExecuteClusterTransactionOneByOne(List<ClusterTransactionCommand.Si
             try
             {
                 var index = command.Index;
+
                 var options = mergedCommands.Options[index];
                 if (exception == null)
                 {
-                    Task indexTask = null;
-                    if (options.WaitForIndexesTimeout != null)
-                    {
-                        indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout.Value,
-                            options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
-                            mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
-                    }
 
-                    var result = new BatchHandler.ClusterTransactionCompletionResult
+                    var indexTask = BatchHandler.WaitForIndexesAsync(DocumentsStorage.ContextPool, this, options.WaitForIndexesTimeout,
+                        options.SpecifiedIndexesQueryString, options.WaitForIndexThrow,
+                        mergedCommands.LastDocumentEtag, mergedCommands.LastTombstoneEtag, mergedCommands.ModifiedCollections);
+
+                    var removeTask = ClusterTransactionWaiter.CreateTask(command.Options.TaskId, command.Index, out _);
+
+                    indexTask.ContinueWith(t =>
                     {
-                        Array = mergedCommands.Replies[index],
-                        IndexTask = indexTask,
-                    };
-                    ClusterTransactionWaiter.SetResult(options.TaskId, index, result);
+                        try
+                        {
+                            t.GetAwaiter().GetResult(); // Task t is already completed here
+                            ClusterTransactionWaiter.SetResult(options.TaskId, index, command.PreviousCount);
+                        }
+                        catch (Exception e)
+                        {
+                            ClusterTransactionWaiter.SetException(options.TaskId, index, e);
+                        }
+
+                        ThreadingHelper.InterlockedExchangeMax(ref LastCompletedClusterTransactionIndex, index);
+                        removeTask.Dispose();
+                    });
+
                     _nextClusterCommand = command.PreviousCount + command.Commands.Length;
                     _lastCompletedClusterTransaction = _nextClusterCommand.Value - 1;
+
+                    ThreadingHelper.InterlockedExchangeMax(ref LastCompletedClusterTransactionIndex, index);
```

Comment:
you are setting this again here, even if indexTask did not complete?

Reply:
the code here is confusing... let's be more explicit about the if we really need to wait for the indexes or not.. (revert to the previous impl. where we had an `if` statement)

Reply:
You are right. the indexTask is a completed task in case there is no need for 'waiting for index', so the '`InterlockedExchangeMax` should be called once inside the `ContinueWith' callback.